src/IMG_libpng.c

```diff
@@ -0,0 +1,2181 @@
+        if (vars->color_ptr == NULL) {
+            vars->error = "Couldn't allocate palette for PNG file";
+            return false;
+        }
+        for (i = 0; i < ncolors; i++) {
+            vars->color_ptr[i].red = palette->colors[i].r;
+            vars->color_ptr[i].green = palette->colors[i].g;
+            vars->color_ptr[i].blue = palette->colors[i].b;
+            if (palette->colors[i].a != 255) {
+                last_transparent = i;
+            }
+        }
+        lib.png_set_PLTE(vars->png_ptr, vars->info_ptr, vars->color_ptr, ncolors);
+        png_color_type = PNG_COLOR_TYPE_PALETTE;
+
+        if (last_transparent >= 0) {
+            for (i = 0; i <= last_transparent; ++i) {
+                transparent_table[i] = palette->colors[i].a;
+            }
+            lib.png_set_tRNS(vars->png_ptr, vars->info_ptr, transparent_table, last_transparent + 1, NULL);
+        }
+    } else if (surface->format == SDL_PIXELFORMAT_RGB24) {
+        png_color_type = PNG_COLOR_TYPE_RGB;
+    } else if (!SDL_ISPIXELFORMAT_ALPHA(surface->format)) {
+        png_color_type = PNG_COLOR_TYPE_RGB;
+        vars->source_surface_for_save = SDL_ConvertSurface(surface, SDL_PIXELFORMAT_RGB24);
+        if (!vars->source_surface_for_save) {
+            vars->error = SDL_GetError();
+            return false;
+        }
+    } else {
+        png_color_type = PNG_COLOR_TYPE_RGBA;
+        vars->source_surface_for_save = SDL_ConvertSurface(surface, SDL_PIXELFORMAT_RGBA32);
+        if (!vars->source_surface_for_save) {
+            vars->error = SDL_GetError();
+            return false;
+        }
+    }
+
+    lib.png_set_IHDR(vars->png_ptr, vars->info_ptr, vars->source_surface_for_save->w, vars->source_surface_for_save->h,
+                     bit_depth, png_color_type, PNG_INTERLACE_NONE,
+                     PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);
+
+    lib.png_write_info(vars->png_ptr, vars->info_ptr);
+
+    vars->row_pointers = (png_bytep *)SDL_malloc(sizeof(png_bytep) * vars->source_surface_for_save->h);
+    if (!vars->row_pointers) {
+        vars->error = "Out of memory allocating row pointers";
+        return false;
+    }
+    for (int row = 0; row < (int)vars->source_surface_for_save->h; row++) {
+        vars->row_pointers[row] = (png_bytep)((Uint8 *)vars->source_surface_for_save->pixels + row * (size_t)vars->source_surface_for_save->pitch);
+    }
+
+    lib.png_write_image(vars->png_ptr, vars->row_pointers);
+    lib.png_write_end(vars->png_ptr, vars->info_ptr);
+
+    return true;
+}
+#endif // SDL_IMAGE_SAVE_PNG
+
+bool IMG_SavePNG_IO(SDL_Surface *surface, SDL_IOStream *dst, bool closeio)
+{
+#if !defined(SDL_IMAGE_SAVE_PNG)
+    return false;
+#else
+    if (!surface || !dst) {
+        SDL_SetError("Surface or SDL_IOStream is NULL");
+        return false;
+    }
+
+    if (!IMG_InitPNG()) {
+        return false;
+    }
+
+    struct png_op_vars vars;
+    bool result = false;
```

Comment:
When dealing with `setjmp()` and `longjmp()` it is probably best if we use a C89-style code layout (all variable declarations at the top of a scope, followed by the rest of the code) to make it easier to audit whether they can be clobbered by the `longjmp()`.